drivers/SmartThings/jbl/src/init.lua

```diff
@@ -0,0 +1,221 @@
+      device:offline()
+    end
+
+    eventsource.onopen = function(msg)
+      log.info("Eventsource open: dni = " .. tostring(device.device_network_id))
+      device:online()
+    end
+
+    local old_eventsource = device:get_field(fields.EVENT_SOURCE)
+    if old_eventsource then
+      log.info("Eventsource Close: dni = " .. tostring(device.device_network_id))
+      old_eventsource:close()
+    end
+    device:set_field(fields.EVENT_SOURCE, eventsource)
+  end
+end
+
+local function update_connection(driver, device, device_ip, device_info)
+  local device_dni = device.device_network_id
+  log.info("update connection, dni = " .. tostring(device_dni))
+
+  local conn_info = driver.discovery_helper.get_connection_info(driver, device_dni, device_ip, device_info)
+
+  local credential = device:get_field(fields.CREDENTIAL)
+
+  conn_info:add_header(CREDENTIAL_KEY_HEADER, credential)
+
+  if driver.device_manager.is_valid_connection(driver, device, conn_info) then
+    device:set_field(fields.CONN_INFO, conn_info)
+
+    create_sse(driver,device, credential)
+  end
+end
+
+
+local function find_new_connetion(driver, device)
+  log.info("find new connection for dni=" .. tostring(device.device_network_id))
+  local ip_table = discovery.find_ip_table(driver)
+  local ip = ip_table[device.device_network_id]
+  if ip then
+    device:set_field(fields.DEVICE_IPV4, ip, {persist = true})
+    local device_info = device:get_field(fields.DEVICE_INFO)
+    update_connection(driver, device, ip, device_info)
+  end
+end
+
+local function check_and_update_connection(driver, device)
+  local conn_info = device:get_field(fields.CONN_INFO)
+  if not driver.device_manager.is_valid_connection(driver, device, conn_info) then
+    device:offline()
+    find_new_connetion(driver, device)
+    conn_info = device:get_field(fields.CONN_INFO)
+  end
+
+  if driver.device_manager.is_valid_connection(driver, device, conn_info) then
+    device:online()
+  end
+end
+
+local function create_monitoring_thread(driver, device, device_info)
+  local old_timer = device:get_field(fields.MONITORING_TIMER)
+  if old_timer ~= nil then
+    log.info("monitoring_timer: dni=" .. device.device_network_id .. ", remove old timer")
+    device.thread:cancel_timer(old_timer)
+  end
+
+  local monitoring_interval = DEFAULT_MONITORING_INTERVAL
+
+  log.info("create_monitoring_thread: dni=" .. device.device_network_id)
+  local new_timer = device.thread:call_on_schedule(monitoring_interval, function()
+    check_and_update_connection(driver, device)
+    driver.device_manager.device_monitor(driver, device, device_info)
+    end, "monitor_timer")
+  device:set_field(fields.MONITORING_TIMER, new_timer)
+end
+
+local function refresh(driver, device, cmd)
+  log.info("refresh : dni =  " .. tostring(device.device_network_id))
+  check_and_update_connection(driver, device)
+  driver.device_manager.refresh(driver, device)
+end
+
+local function device_removed(driver, device)
```

Comment:
Since you are already capturing the spawned timers in fields, it would be beneficial to tear those down on removal as well.

Reply:
@cchillerr

Reply:
I dont see any timers being cancelled.

Reply:
![image](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/assets/138542568/b7744598-d0a2-4e7b-8d99-f282b6c84b22)
https://github.com/cchillerr/SmartThingsEdgeDrivers/blob/main/drivers/SmartThings/jbl/src/init.lua 
It seems the code is not up to date here?

Reply:
@cjswedes Sorry, I'm confused, which spawned timers we are talking about here?

Reply:
@cjswedes @cchillerr I believe that with some of the refactors that were made, this is no longer needed.